### PR TITLE
Serial drivers rework 2

### PR DESCRIPTION
**Pull request by jforissier**
Replaces https://github.com/OP-TEE/optee_os/pull/1393. First review comments are hopefully addressed. What has changed:
- Dropped the double initialization with PA then VA with `console_init_early()`/`console_init()`. Instead, I have introduced `phys_to_virt_io()` which converts a PA to a VA by using a `MEM_AREA_IO_SEC` mapping if it exists, `MEM_AREA_IO_NSEC` otherwise. With this, the UART drivers can use common code for PA to VA translation when needed (the new `struct io_pava` and `io_pa_or_va()` helper make this even easier).
- I have introduced `register_serial_console()`
- I have removed `core/arch/arm/plat-ti/console.c` which was left with only one function: `console_init()`.

### CODE REVIEW

File: core/drivers/sunxi_uart.c [hunk 1]
```diff
@@ -86,12 +102,19 @@ void sunxi_uart_putc(int ch, vaddr_t base)
...
-	while (!sunxi_uart_have_rx_data(base)) {
-		/* transmit fifo is empty, waiting again. */
-		;
-	}
-	return read32(base + UART_REG_RBR) & 0xff;
-}
+	pd->base.pa = base;
+	pd->chip.ops = &sunxi_uart_ops;
 
+	/* do nothing, debug uart(uart0) share with normal world,
```
Comment (etienne-lms) [inline core/drivers/sunxi_uart.c]: minor: newline after `/*`

File: core/arch/arm/plat-sprd/console.c [hunk 2]
```diff
@@ -43,9 +44,3 @@ void console_putc(int ch)
 	cons->ops->putc(cons, ch & 0xff);
 }
```
Comment (etienne-lms) [inline core/arch/arm/plat-sprd/console.c]: Could you add a comment that this platform specific function differs from the generic one as `\n` ar not pre-fixed with a `\r`? Maybe the platform maintainer will want to resync.

File: core/arch/arm/plat-sunxi/console.c [hunk 3]
```diff
@@ -35,6 +35,7 @@ static struct sunxi_uart_data console_data __early_bss;
 void console_init(void)
 {
 	sunxi_uart_init(&console_data, CONSOLE_UART_BASE);
+	register_serial_console(&console_data.chip);
 }
 
 void console_putc(int ch)
```
Comment (etienne-lms) [inline core/arch/arm/plat-sunxi/console.c]: Same comment as for **plat-sprd/console.c**

File: core/drivers/hi16xx_uart.c [hunk 4]
```diff
@@ -76,16 +79,69 @@
 #define UART_USR_RFNE_BIT	3	/* Receive FIFO not empty bit */
 #define UART_USR_RFF_BIT	4	/* Receive FIFO full bit */
 
-void hi16xx_uart_flush(vaddr_t base)
+static void hi16xx_uart_flush(struct serial_chip *chip)
 {
+	struct hi16xx_uart_data *pd = container_of(chip,
```
Comment (jenswi-linaro) [inline core/drivers/hi16xx_uart.c]: All these functions would be a bit less complicated with a helper function like:
```
static vaddr_t chip_to_base(struct serial_chip *chip)
{
        struct hi16xx_uart_data *pd =
                container_of(chip, struct hi16xx_uart_data, chip);

        return io_pa_or_va(&pd->base);
}
```

File: core/kernel/console.c [hunk 5]
```diff
@@ -26,42 +26,31 @@
...
-#include <mm/core_memprot.h>
-#include <platform_config.h>
+#include <compiler.h>
+#include <drivers/serial.h>
+#include <stdlib.h>
 
-register_phys_mem(MEM_AREA_IO_NSEC,
-		  CONSOLE_UART_BASE,
-		  SERIAL8250_UART_REG_SIZE);
+struct serial_chip *serial_console __early_bss;
```
Comment (jenswi-linaro) [inline core/kernel/console.c]: This doesn't seem to be used outside this file, so it could be static.

File: core/drivers/hi16xx_uart.c (hunk 4 above)
Comment (jforissier) [inline core/drivers/hi16xx_uart.c]: Yes obviously. I'll update the code.

File: core/drivers/sunxi_uart.c (hunk 1 above)
Comment (jforissier) [inline core/drivers/sunxi_uart.c]: Will fix

File: core/include/drivers/serial.h [hunk 6]
```diff
@@ -40,4 +44,25 @@ struct serial_ops {
 	int (*getchar)(struct serial_chip *chip);
 };
 
+struct io_pava {
+	paddr_t pa;
+	vaddr_t va;
+};
```
Comment (jforissier) [inline core/include/drivers/serial.h]: True, but I did not think it was worth creating a new file just for this. Where would you put this code? It's only used in serial drivers currently, I thought we'll move it if/when we want to use it elsewhere.

OK for `s/io_pava/io_pa_va`.

File: core/include/drivers/serial8250_uart.h [hunk 7]
```diff
@@ -28,19 +28,18 @@
...
-void serial8250_uart_putc(int ch, vaddr_t base);
-
-void serial8250_uart_flush_tx_fifo(vaddr_t base);
-
-bool serial8250_uart_have_rx_data(vaddr_t base);
-
-int serial8250_uart_getchar(vaddr_t base);
+void serial8250_uart_init(struct serial8250_uart_data *pd, paddr_t base,
+			  uint32_t __unused uart_clk,
+			  uint32_t __unused baud_rate);
```
Comment (jforissier) [inline core/include/drivers/serial8250_uart.h]: True, I'll fix.
That's interesting. I think my editor (I use Qt Creator when I don't use vi) has propagated the `__unused` to the prototype as I changed `vaddr_t` to `paddr_t` for `base`.

File: core/kernel/console.c (hunk 5 above)
Comment (jforissier) [inline core/kernel/console.c]: Yes, it should. Will fix.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Pull request merged**